p4-16/spec/P4-16-spec.mdk

```diff
@@ -5523,6 +5524,37 @@ use to implement generalized switch statements.
 
 [GeneralizedSwitchStatements]: <https://github.com/p4lang/p4-spec/blob/master/p4-16/spec/docs/implementing-generalized-switch-statements.md>
 
+## Loop statement { #sec-loop-stmt }
+
+A `loop` statement executes a statement zero or more times, based on a condition
+or a range or collection.
+
+~ Begin P4Grammar
+
+[INCLUDE=grammar.mdk:loopStatement]
+
+[INCLUDE=grammar.mdk:forInitStatement]
+
+[INCLUDE=grammar.mdk:forUpdateStatement]
+
+~ End P4Grammar
+
+The basic 3-clause `for` statement is similar to a C for statement.  The init statements
+will be executed prior to executing the loop.  The condition will be evaluated before
+the first iteration and after each iteration, to determine if the loop should exit.
+The update statements will be executed after the loop body and before evaluating the
+condition again for the next iteration.
+
+The `for`-`in` statement executes the body once for each value in a range or each
+element in a list expression or header stack.
```

Comment:
Maybe point out that the range endpoints, and list elements, are only evaluated once before the first execution of the loop body?

Reply:
This brings up a potentially annoying corner case -- for something like `for T x in (list<T>){ a, b(arg), c(arg) } body` where there are side effects within the list expression, but assoicated with specific elements, must all side effects in those function calls occur before the first iteration?  A simple unrolling might want to evaluate each list element initializer just before that element's iteration.

Probably be best to require evaluating all the side effects first.  After unrolling, copy propagation should be able to move down side effects if that is legal and useful.

Reply:
The only two alternatives I can think of that might make sense are:

(a) specify that all evaluation (including side effects, if any) of evaluating list elements is done before the first loop iteration
(b) specify that the evaluation of a list element is done just before the iteration that uses that list element, which might have changed since the first loop iteration.

I'm not dead-set against (b).  I realize that (a) could require a lot of temporary storage of list elements in the worst case, but we can explain to developers this fact, and they can avoid such programs that require that if they prefer.